### PR TITLE
Support updates when Delta table location has two trailing slashes

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Alternative approach to https://github.com/trinodb/trino/pull/18155 for obtaining a sibling file location.

Fixes https://github.com/trinodb/trino/issues/17966

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OriginalFilesUtils.java [hunk 1]
```diff
@@ -54,7 +54,7 @@ public static long getPrecedingRowCount(
         long rowCount = 0;
         for (OriginalFileInfo originalFileInfo : originalFileInfos) {
             if (originalFileInfo.getName().compareTo(splitPath.fileName()) < 0) {
-                Location path = splitPath.parentDirectory().appendPath(originalFileInfo.getName());
+                Location path = splitPath.replaceFilename(originalFileInfo.getName());
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OriginalFilesUtils.java:57]: Let's move this to separate commit, so that Delta fix comes with its test changes only and other cleanup (like this one + same in  `io.trino.plugin.deltalake.transactionlog.writer.S3NativeTransactionLogSynchronizer#write` and `HiveWriterFactory`) come as a second commit.

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java [hunk 2]
```diff
@@ -178,6 +178,17 @@ public String fileName()
...
 
+    /**
+     * Returns a new location with the same parent directory as the current location,
+     * but with the filename corresponding to the specified `filename`.
+     * The location must be a valid file location.
+     */
+    public Location replaceFilename(String filename)
+    {
+        verifyValidFileLocation();
+        return Location.of(location.replaceAll("[^/]+$", "") + filename);
```
Comment (findepi) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java:189]: - use lastIndexOf("/") instead of regular expressions to be much faster 
- use `io.trino.filesystem.Location#withPath` to avoid re-parsing the location

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OriginalFilesUtils.java (hunk 1 above)
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OriginalFilesUtils.java:57]: Follow-up `HiveWriterFactory` https://github.com/trinodb/trino/pull/18183

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java [hunk 3]
```diff
@@ -178,6 +178,21 @@ public String fileName()
...
+    /**
+     * Returns a new location with the same parent directory as the current location,
+     * but with the filename corresponding to the specified name.
+     * The location must be a valid file location.
+     */
+    public Location sibling(String name)
+    {
+        requireNonNull(name, "name is null");
+        checkArgument(!name.isEmpty(), "name is empty");
+        checkArgument(!name.isBlank(), "name is blank");
```
Comment (findepi) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java:190]: isBlank has no special meaning here.
we either support leading/trailing whitespace or not, regardless whether there is any non-whitespace character in the middle

we probably don't need to support `.sibling(" ")` , but from filesystem location design perspective, this should be valid.

see https://github.com/trinodb/trino/pull/18206

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OriginalFilesUtils.java [hunk 4]
```diff
@@ -54,7 +54,7 @@ public static long getPrecedingRowCount(
         long rowCount = 0;
         for (OriginalFileInfo originalFileInfo : originalFileInfos) {
             if (originalFileInfo.getName().compareTo(splitPath.fileName()) < 0) {
-                Location path = splitPath.parentDirectory().appendPath(originalFileInfo.getName());
+                Location path = splitPath.sibling(originalFileInfo.getName());
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OriginalFilesUtils.java:57]: https://github.com/trinodb/trino/pull/18177#discussion_r1258034677

**Review by findepi — APPROVED** [review]
